kernel/ptrace.c

```diff
@@ -0,0 +1,243 @@
+#include "ptrace.h"
+#include "kernel/calls.h"
+#include "kernel/errno.h"
+#include "kernel/signal.h"
+#include "task.h"
+#include <string.h>
+
+// Returns stopped child with the given pid, locked with the ptrace lock
+static struct task *find_child(pid_t_ pid) {
+    struct task *child = NULL;
+    list_for_each_entry(&current->children, child, siblings) {
+        if (child->pid == pid) {
+            lock(&child->ptrace.lock);
+            if (child->ptrace.stopped) {
+                break;
+            } else {
+                unlock(&child->ptrace.lock);
+            }
+        }
+    }
+    return child;
```

Comment:
Just noticed what I think is a bug here: if none of the children have the given PID, it will return the last child, not NULL.